### ISSUE TITLE
debugging function can't work with password stored in .pgpass file

### ISSUE THREAD

**Issue by dturon** [OmniDB/OmniDB#432]
Hi,

when password is stored in .pgpass file OmniDB can connect to database, but when i try debug some function OmniDB send empty password.
```
2018-03-23 10:14:20.926 CET,"dbuser","dbname",27932,"10.76.3.51:44426",5ab4c56c.6d1c,9,"SELECT",2018-03-23 10:14:20 CET,2/3071,0,LOG,00000,"omnidb, CONNINFO, (host='dbhost' port=5432 dbname='dbname' user='dbuser' password='' application_name='OmniDB')",,,,,"PL/pgSQL function schema.fn(text,integer) during function entry","select * from schema.fn('test',5000) limit 1000",,,"OmniDB"
2018-03-23 10:14:20.927 CET,,,27934,"10.1.1.1:38526",5ab4c56c.6d1e,1,"",2018-03-23 10:14:20 CET,,0,LOG,00000,"connection received: host=10.1.1.1 port=38526",,,,,,,,,""
2018-03-23 10:14:20.928 CET,"dbuser","dbname",27932,"10.76.3.51:44426",5ab4c56c.6d1c,10,"SELECT",2018-03-23 10:14:20 CET,2/3071,0,ERROR,XX000,"omnidb: Connection to database failed: fe_sendauth: no password supplied
```

other issue is storing passwords in LOG file..., but on devel server isn't big issue.

Thanks and have a nice day!

David

**Comment by rafaelthca** [comment]
Hi @dturon,

Noted. Debugger is being improved for the next release (that is coming out in around 3 weeks). Passwords won't be logged anymore and we are changing the way connections are opened.

In your case, please note that the debugger opens a new connection that updates control tables for OmniDB to get the current state of the debugging operation. This connection is opened by the session that is running the function being debugged, so in this case the OS user that creates the connection is postgres. As a consequence you need to include the same line in .pgpass of the postgres user.

As I mentioned before, this authentication of the third connection is being improved for the next release. The plugin will also be available for PostgreSQL running on Windows.

Rafael.

**Comment by dturon** [comment]
Hi Rafael,

thanks I am looking forward for new release. Logged user was't postgres ... i replace name to dbuser for security reasons. Name was same as declared in connection string in OmniDB app. If we did not use .pgpass file and filled password in connection string debuging works.

```
2018-03-23 10:19:42.292 CET,"dbuser","dbname",979,"10.76.3.51:47452",5ab4c6ae.3d3,8,"SELECT",2018-03-23 10:19:42 CET,18/0,0,LOG,00000,"duration: 0.215 ms  statement: select omnidb.omnidb_enable_debugger('host=''dbhost'' port=5432 dbname=''dbname'' user=''dbuser'' password='''' application_name=''OmniDB''')",,,,,,,,,"OmniDB"
```  

Have a nice day.
David

**Comment by dturon** [comment]
Ok now i understand. Thanks

**Comment by wind39** [comment]
Hi @dturon ,

Can you please test latest release and provide your valuable feedback? Thanks very much in advance!

Regards,
William Ivanski

**Comment by dturon** [comment]
Hi,

i loaded test functions but click on debug function in GUI:
`
2018-07-09 15:25:29.713 CEST,"test_omni","postgres",25679,"10.76.3.24:43964",5b436249.644f,3,"SELECT",2018-07-09 15:25:29 CEST,2/8,0,ERROR,XX000,"failed to find conversion function from unknown to regprocedure",,,,,,"
            with obj as (
                SELECT p.oid,
                       p.proname AS name,
                       n.nspname AS namespace,
                       pg_get_userbyid(p.proowner) AS owner,
                       'omnidb.function_01(text, text)' AS sql_identifier
                FROM pg_proc p JOIN pg_namespace n ON n.oid=p.pronamespace
                WHERE p.oid = 'omnidb.function_01(text, text)'::regprocedure
            ),
            createfunction as (
                select pg_get_functiondef(sql_identifier::regprocedure)||E'
' as text
                from obj
            ),
            alterowner as (
                select
                   'ALTER FUNCTION '||sql_identifier||
                          ' OWNER TO '||quote_ident(owner)||E';
                  
' as text
                  from obj
            ),
            privileges as (
            select (u_grantor.rolname)::information_schema.sql_identifier as grantor,
                   (grantee.rolname)::information_schema.sql_identifier as grantee,
                   (p.privilege_type)::information_schema.character_data as privilege_type,
                   (case when (pg_has_role(grantee.oid, p.proowner, 'USAGE'::text) or p.is_grantable)
                         then 'YES'::text
                         else 'NO'::text
                    end)::information_schema.yes_or_no AS is_grantable
            from (
                select p.pronamespace,
                       p.proowner,
                       (aclexplode(COALESCE(p.proacl, acldefault('f', p.proowner)))).grantor as grantor,
                       (aclexplode(COALESCE(p.proacl, acldefault('f', p.proowner)))).grantee as grantee,
                       (aclexplode(COALESCE(p.proacl, acldefault('f', p.proowner)))).privilege_type as privilege_type,
                       (aclexplode(COALESCE(p.proacl, acldefault('f', p.proowner)))).is_grantable as is_grantable
                from pg_proc p
                where not p.proisagg
                  and p.oid = 'omnidb.function_01(text, text)'::regprocedure
            ) p
            inner join pg_namespace n
            on n.oid = p.pronamespace
            inner join pg_roles u_grantor
            on u_grantor.oid = p.grantor
            inner join (
                select r.oid,
                       r.rolname
                from pg_roles r
                union all
                select (0)::oid AS oid,
                       'PUBLIC'::name
            ) grantee
            on grantee.oid = p.grantee
            ),
            grants as (
            select coalesce(
                    string_agg(format(
                        E'GRANT %s ON FUNCTION omnidb.function_01(text, text) TO %s%s;
',
                    privilege_type,
                    case grantee
                      when 'PUBLIC' then 'PUBLIC'
                      else quote_ident(grantee)
                    end,
                        case is_grantable
                      when 'YES' then ' WITH GRANT OPTION'
                      else ''
                    end), ''),
                   '') as text
            from privileges
            )
            select (select text from createfunction) ||
                   (select text from alterowner) ||
                   (select text from grants)
",,,"OmniDB"

`
tested on centos6, postgres 9.6

**Comment by wind39** [comment]
Hi @dturon ,

This is the same issue as #591 , and we have fixed in OmniDB 2.10. Can you please try again and give us your valuable feedback?

Thanks in advance.

William Ivanski

**Comment by wind39** [comment]
Closing this now.
